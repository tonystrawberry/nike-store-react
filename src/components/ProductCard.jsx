import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { unselectProduct } from '../redux/actions';


import './ProductOverview.css';
import './ProductCard.css';
import Logo from '../assets/nike_logo.svg';
import NikeAirForceSmall from '../assets/nike_airforce-small.jpg';
import NikeAirForceMedium from '../assets/nike_airforce-medium.jpg';
import NikeAirForceLarge from '../assets/nike_airforce-large.jpg';


const mapStateToProps = (state) => {
  return {

  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onUnselected: () => dispatch(unselectProduct())
  }
}

class ProductCard extends PureComponent {
  constructor(props){
    super(props);

    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
    window.addEventListener('resize', this.updateDimensions);
    this.updateDimensions();
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
    window.removeEventListener('resize', this.updateDimensions);
  }

  /**
   * Set the wrapper ref
   */
  setWrapperRef(node) {
    this.wrapperRef = node;
  }

  /**
   * Alert if clicked on outside of element
   */
  handleClickOutside(event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.props.onUnselected();
    }
  }

  updateDimensions() {
    let x = window.matchMedia("(max-width: 1000px)");
    const photo = document.querySelectorAll('.product-card__photo');
    const pc = document.querySelector('.product-card__photo-container');
  
    if (x.matches){
      // shoeBackgroundの新しいheight計算
      let width = photo[0].offsetWidth; // offsetHeight = elementのheight + elementのpadding + elementのborder
      pc.style.height = `${width}px`;
    } else {
      pc.style.height = '475px';
    }
  }

  render() {
    return (
      <div ref={this.setWrapperRef} className="product-card__container">
        <div className="product-card__card">
          <div className="product-card__photo-container">
            <img src={Logo} alt="Nike Logo" className="product-card__logo" />
            <a href="#" className="product-card__share"><i className="fas fa-share-alt"></i></a>
            <img src={NikeAirForceLarge} 
              srcSet={`${NikeAirForceSmall} 216w, ${NikeAirForceMedium} 432w, ${NikeAirForceLarge} 864w`} 
              sizes='(max-width: 500px) 432px, 864px'
              alt="Product Photography" className="product-card__photo"/>
          </div>
          <div className="product-card__info">
            <div className="product-card__name">
              <div>
                <h1 className="big">Nike Air Force 1</h1>
                {this.props.new ? <span className="new">new</span> : null}
              </div>
              <h3 className="small">Men's Shoe</h3>
            </div>
            <div className="product-card__description">
              <p className="product-card__text">The legend lives on in the Nike Air Force 1, which stays true to its roots with iconic AF1 style and Nike Air for all-day comfort and long-lasting wear.</p>
            </div>
            <div className="product-card__color-container">
              <h3 className="product-card__title">Color</h3>
              <div className="product-card__colors">
                <span className="product-card__color" primary="#2175f5" color="blue"></span>
                <span className="product-card__color" primary="#f84848" color="red"></span>
                <span className="product-card__color" primary="#29b864" color="green"></span>
                <span className="product-card__color" primary="#ff5521" color="orange"></span>
                <span className="product-card__color" primary="#444" color="black"></span>
              </div>
            </div>
            <div className="product-card__size-container">
              <h3 className="product-card__title">size</h3>
              <div className="product-card__sizes">
                <span className="product-card__size">7</span>
                <span className="product-card__size">8</span>
                <span className="product-card__size">9</span>
                <span className="product-card__size">10</span>
                <span className="product-card__size">11</span>
              </div>
            </div>
            <div className="product-card__buy-price">
              <a href="#" className="product-card__buy"><i className="fas fa-shopping-cart"></i>Add to cart</a>
              <div className="product-card__price">
                <i className="fas fa-yen-sign"></i>
                <h1>21,000</h1>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard);