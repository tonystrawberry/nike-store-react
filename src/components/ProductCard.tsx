import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { unselectProduct } from '../redux/actions';
import { Dispatch } from 'redux';

import './ProductOverview.scss';
import './ProductCard.scss';
import Logo from '../assets/nike_logo.svg';
import { Product } from '../types';
import Loading from './Loading';
import { numberToPrice } from '../utils/utils';


const mapStateToProps = () => {
  return {

  }
}

const mapDispatchToProps = (dispatch : Dispatch ) => {
  return {
    onUnselected: () => dispatch(unselectProduct())
  }
}

interface IProductCardProps {
  new: boolean,
  onUnselected: {(): void},
  product: Product | undefined
}

interface IProductCardState {
  
}


class ProductCard extends PureComponent<IProductCardProps, IProductCardState> {
  constructor(props : IProductCardProps){
    super(props);

    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this.wrapperRef = null;
  }

  wrapperRef: HTMLDivElement | null;

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
    window.addEventListener('resize', this.updateDimensions);
    this.updateDimensions();
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
    window.removeEventListener('resize', this.updateDimensions);
  }

  /**
   * Set the wrapper ref
   */
  setWrapperRef(node : HTMLDivElement | null) : void{
    this.wrapperRef= node;
  }

  /**
   * Alert if clicked on outside of element
   */
  handleClickOutside(event : Event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target as Node)) {
      this.props.onUnselected();
    }
  }

  updateDimensions() {
    let x = window.matchMedia("(max-width: 1000px)");
    const photo = document.querySelectorAll('.product-card__photo') as NodeListOf<HTMLElement>;
    const pc = document.querySelector('.product-card__photo-container') as HTMLElement;
  
    if (x.matches){
      // shoeBackgroundの新しいheight計算
      let width = photo[0].offsetWidth; // offsetHeight = elementのheight + elementのpadding + elementのborder
      pc.style.height = `${width}px`;
    } else {
      pc.style.height = '475px';
    }
  }

  render() {
    if (!this.props.product){
      return (
        <div className="product-card__container">
          <Loading />
        </div>)
    }
    return (
      <div ref={this.setWrapperRef} className="product-card__container">
        <div className="product-card__card">
          <div className="product-card__photo-container">
            <img src={Logo} alt="Nike Logo" className="product-card__logo" />
            <a href="/" className="product-card__share"><i className="fas fa-share-alt"></i></a>
            <img src={this.props.product.imageUrl} 
              alt="Product Photography" className="product-card__photo"/>
          </div>
          <div className="product-card__info">
            <div className="product-card__name">
              <div>
                <h1 className="big">{ this.props.product.title }</h1>
                {this.props.new ? <span className="new">new</span> : null}
              </div>
              <h3 className="small">{ this.props.product.subtitle1 }</h3>
              <h3 className="small">{ this.props.product.subtitle2 }</h3>
            </div>
            <div className="product-card__description">
              <p className="product-card__text">{ this.props.product.description }</p>
            </div>
            {/* <div className="product-card__color-container">
              <h3 className="product-card__title">Color</h3>
              <div className="product-card__colors">
                <span className="product-card__color" color="blue"></span>
                <span className="product-card__color" color="red"></span>
                <span className="product-card__color" color="green"></span>
                <span className="product-card__color" color="orange"></span>
                <span className="product-card__color" color="black"></span>
              </div>
            </div>
            <div className="product-card__size-container">
              <h3 className="product-card__title">size</h3>
              <div className="product-card__sizes">
                <span className="product-card__size">7</span>
                <span className="product-card__size">8</span>
                <span className="product-card__size">9</span>
                <span className="product-card__size">10</span>
                <span className="product-card__size">11</span>
              </div>
            </div> */}
            <div className="product-card__buy-price">
              <a href="/" className="product-card__buy"><i className="fas fa-shopping-cart"></i>Add to cart</a>
              <div className="product-card__price">
                <i className="fas fa-yen-sign"></i>
                <h1>{ numberToPrice(this.props.product.price) }</h1>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard);