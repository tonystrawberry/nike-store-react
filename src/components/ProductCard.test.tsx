import { mount, shallow, ReactWrapper } from 'enzyme';
import ProductCard from './ProductCard';
import { Provider } from 'react-redux';
import React, { Suspense } from 'react';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';

const mockStore = configureStore([]);

beforeEach(() => {
  Object.defineProperty(document, 'querySelectorAll', {
    writable: true,
    value: jest.fn().mockImplementation(query => ([{
      offsetWidth: 400
    }])),
  })

  Object.defineProperty(document, 'querySelector', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      style: {
        height: 0
      }
    })),
  })

  Object.defineProperty(document, 'querySelector', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      style: {
        height: 0
      }
    })),
  })
})

it('should call removeEventListener in componentWillUnmount', () => {

  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: true,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  const documentRemoveListener = jest.spyOn(document, 'removeEventListener');
  const windowRemoveListener = jest.spyOn(window, 'removeEventListener');
  
  let store = mockStore({});
  let wrapper = mount(<Provider store={store}><ProductCard new={true} /></Provider>);
  wrapper.unmount(); // Calls for componentWillUnmount

  expect(documentRemoveListener).toBeCalled();
  expect(windowRemoveListener).toBeCalled();
});


it ('expect to render ProductCard component with matchMedia => true', () => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: true,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  let store = mockStore({});
  let wrapper = mount(<Provider store={store}><ProductCard new={true} /></Provider>);

  expect(toJson(wrapper)).toMatchSnapshot();
})

it ('expect to render ProductCard component with matchMedia => false', () => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  let store = mockStore({});
  let wrapper = mount(<Provider store={store}><ProductCard new={true} /></Provider>);

  expect(toJson(wrapper)).toMatchSnapshot();
})

it ('handle clicks outside the component', () => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  let wrapper : ReactWrapper <any, any>;

  const mapDocument : any = {};
  document.addEventListener = jest.fn((event, cb : {( target : object ): void}) => {
    mapDocument[event] = () => {
      cb({ target: wrapper.find(".outside").last().getDOMNode() })
    };
  });  

  const documentAddListener = jest.spyOn(document, 'addEventListener');

  let store = mockStore({});
  wrapper = mount(
    <div> 
      <Provider store={store}><ProductCard new={true} /></Provider>
      <div className="outside"></div>
    </div>
  );

  expect(documentAddListener).toBeCalled();

  mapDocument.mousedown()
})

it ('handle clicks inside the component', () => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  let wrapper : ReactWrapper <any, any>;

  const mapDocument : any = {};
  document.addEventListener = jest.fn((event, cb : {( target : object ): void}) => {
    mapDocument[event] = () => {
      cb({ target: wrapper.find(".product-card__container").last().getDOMNode() })
    };
  });  

  const documentAddListener = jest.spyOn(document, 'addEventListener');

  let store = mockStore({});
  wrapper = mount(
    <div> 
      <Provider store={store}><ProductCard new={true} /></Provider>
      <div className="outside"></div>
    </div>
  );

  expect(documentAddListener).toBeCalled();

  mapDocument['mousedown']()
})

it ('should show new class on new = true props', () => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  let store = mockStore({});
  let wrapper = mount(<Provider store={store}><ProductCard new={true} /></Provider>);

  expect(wrapper.find('.new').length).toBe(1)
})

it ('should not show new class on new = false props', () => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  let store = mockStore({});
  let wrapper = mount(<Provider store={store}><ProductCard new={false} /></Provider>);

  expect(wrapper.find('.new').length).toBe(0)
})


