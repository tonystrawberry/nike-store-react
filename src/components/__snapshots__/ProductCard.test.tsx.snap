// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render ProductCard component with matchMedia => false 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ProductCard)
    new={true}
  >
    <ProductCard
      new={true}
      onUnselected={[Function]}
    >
      <div
        className="product-card__container"
      >
        <div
          className="product-card__card"
        >
          <div
            className="product-card__photo-container"
          >
            <img
              alt="Nike Logo"
              className="product-card__logo"
              src="nike_logo.svg"
            />
            <a
              className="product-card__share"
              href="#"
            >
              <i
                className="fas fa-share-alt"
              />
            </a>
            <img
              alt="Product Photography"
              className="product-card__photo"
              sizes="(max-width: 500px) 432px, 864px"
              src="nike_airforce-medium.jpg"
              srcSet="nike_airforce-medium.jpg 432w, nike_airforce-large.jpg 864w"
            />
          </div>
          <div
            className="product-card__info"
          >
            <div
              className="product-card__name"
            >
              <div>
                <h1
                  className="big"
                >
                  Nike Air Force 1
                </h1>
                <span
                  className="new"
                >
                  new
                </span>
              </div>
              <h3
                className="small"
              >
                Men's Shoe
              </h3>
            </div>
            <div
              className="product-card__description"
            >
              <p
                className="product-card__text"
              >
                The legend lives on in the Nike Air Force 1, which stays true to its roots with iconic AF1 style and Nike Air for all-day comfort and long-lasting wear.
              </p>
            </div>
            <div
              className="product-card__color-container"
            >
              <h3
                className="product-card__title"
              >
                Color
              </h3>
              <div
                className="product-card__colors"
              >
                <span
                  className="product-card__color"
                  color="blue"
                />
                <span
                  className="product-card__color"
                  color="red"
                />
                <span
                  className="product-card__color"
                  color="green"
                />
                <span
                  className="product-card__color"
                  color="orange"
                />
                <span
                  className="product-card__color"
                  color="black"
                />
              </div>
            </div>
            <div
              className="product-card__size-container"
            >
              <h3
                className="product-card__title"
              >
                size
              </h3>
              <div
                className="product-card__sizes"
              >
                <span
                  className="product-card__size"
                >
                  7
                </span>
                <span
                  className="product-card__size"
                >
                  8
                </span>
                <span
                  className="product-card__size"
                >
                  9
                </span>
                <span
                  className="product-card__size"
                >
                  10
                </span>
                <span
                  className="product-card__size"
                >
                  11
                </span>
              </div>
            </div>
            <div
              className="product-card__buy-price"
            >
              <a
                className="product-card__buy"
                href="#"
              >
                <i
                  className="fas fa-shopping-cart"
                />
                Add to cart
              </a>
              <div
                className="product-card__price"
              >
                <i
                  className="fas fa-yen-sign"
                />
                <h1>
                  21,000
                </h1>
              </div>
            </div>
          </div>
        </div>
      </div>
    </ProductCard>
  </Connect(ProductCard)>
</Provider>
`;

exports[`expect to render ProductCard component with matchMedia => true 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ProductCard)
    new={true}
  >
    <ProductCard
      new={true}
      onUnselected={[Function]}
    >
      <div
        className="product-card__container"
      >
        <div
          className="product-card__card"
        >
          <div
            className="product-card__photo-container"
          >
            <img
              alt="Nike Logo"
              className="product-card__logo"
              src="nike_logo.svg"
            />
            <a
              className="product-card__share"
              href="#"
            >
              <i
                className="fas fa-share-alt"
              />
            </a>
            <img
              alt="Product Photography"
              className="product-card__photo"
              sizes="(max-width: 500px) 432px, 864px"
              src="nike_airforce-medium.jpg"
              srcSet="nike_airforce-medium.jpg 432w, nike_airforce-large.jpg 864w"
            />
          </div>
          <div
            className="product-card__info"
          >
            <div
              className="product-card__name"
            >
              <div>
                <h1
                  className="big"
                >
                  Nike Air Force 1
                </h1>
                <span
                  className="new"
                >
                  new
                </span>
              </div>
              <h3
                className="small"
              >
                Men's Shoe
              </h3>
            </div>
            <div
              className="product-card__description"
            >
              <p
                className="product-card__text"
              >
                The legend lives on in the Nike Air Force 1, which stays true to its roots with iconic AF1 style and Nike Air for all-day comfort and long-lasting wear.
              </p>
            </div>
            <div
              className="product-card__color-container"
            >
              <h3
                className="product-card__title"
              >
                Color
              </h3>
              <div
                className="product-card__colors"
              >
                <span
                  className="product-card__color"
                  color="blue"
                />
                <span
                  className="product-card__color"
                  color="red"
                />
                <span
                  className="product-card__color"
                  color="green"
                />
                <span
                  className="product-card__color"
                  color="orange"
                />
                <span
                  className="product-card__color"
                  color="black"
                />
              </div>
            </div>
            <div
              className="product-card__size-container"
            >
              <h3
                className="product-card__title"
              >
                size
              </h3>
              <div
                className="product-card__sizes"
              >
                <span
                  className="product-card__size"
                >
                  7
                </span>
                <span
                  className="product-card__size"
                >
                  8
                </span>
                <span
                  className="product-card__size"
                >
                  9
                </span>
                <span
                  className="product-card__size"
                >
                  10
                </span>
                <span
                  className="product-card__size"
                >
                  11
                </span>
              </div>
            </div>
            <div
              className="product-card__buy-price"
            >
              <a
                className="product-card__buy"
                href="#"
              >
                <i
                  className="fas fa-shopping-cart"
                />
                Add to cart
              </a>
              <div
                className="product-card__price"
              >
                <i
                  className="fas fa-yen-sign"
                />
                <h1>
                  21,000
                </h1>
              </div>
            </div>
          </div>
        </div>
      </div>
    </ProductCard>
  </Connect(ProductCard)>
</Provider>
`;
